File: src/main/java/katas/java/primes/SumOfPrimes.java
package katas.java.primes;

import org.junit.Assert;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.equalTo;

// rbs techstock 2013
public class SumOfPrimes {
    @Test
    public void shouldFindSumOfPrimesBelow2M() {
        Assert.assertThat(sumOfPrimes(100), equalTo(1060L));
        Assert.assertThat(sumOfPrimes(1000000), equalTo(37550402023L)); // ~1 min
        Assert.assertThat(sumOfPrimes(2000000), equalTo(142913828922L)); // ~3 min 40 sec
    }

    private static long sumOfPrimes(int amount) {
        byte prime = 1;
        byte notPrime = 2;
        byte[] numbers = new byte[amount];

        long sum = 0;
        for (int i = 2; i < numbers.length; i++) {
            if (numbers[i] == notPrime) continue;

            boolean isPrime = true;
            for (int j = i; j > 0; j--) {
                if (numbers[j] == prime && i % j == 0) {
                    isPrime = false;
                    break;
                }
            }

            numbers[i] = (isPrime ? prime : notPrime);
            if (isPrime) {
                for (int j = i + i; j < numbers.length; j += i) {
                    numbers[j] = notPrime;
                }
                sum = sum + i;
            }
        }
        return sum;
    }
}

File: scala/src/katas/scala/techstock/SumOfPrimes.scala
package katas.scala.techstock

import org.specs2.matcher.ShouldMatchers
import org.junit.Test
import scala.annotation.tailrec

// rbs techstock 2013
class SumOfPrimes extends ShouldMatchers {
	@Test def `find sum of primes below 2 million`() {
		// too slow for 2M
		def from(n: Int): Stream[Int] = Stream.cons(n, from(n + 1))
		def sieve(s: Stream[Int]): Stream[Int] =
			Stream.cons(s.head, sieve(s.tail filter { _ % s.head != 0}))
		def primes = sieve(from(2))
		primes.take(100).sum should equalTo(24133)

		// too slow for 2M
		@tailrec def primes2(amount: Int = 100, n: Int = 2, result: Seq[Int] = Seq()): Seq[Int] = {
			if (n >= amount) result
			else if (result.isEmpty || result.forall{ n % _ != 0 }) primes2(amount, n + 1, n +: result)
			else primes2(amount, n + 1, result)
		}
		primes2(100).sum should equalTo(1060)
	}
}